internal/fwserver/server_readresource.go

```diff
@@ -71,10 +72,20 @@ func (s *Server) ReadResource(ctx context.Context, req *ReadResourceRequest, res
 		readReq.ProviderMeta = *req.ProviderMeta
 	}
 
+	if req.Private != nil {
+		readReq.Private = req.Private.Provider
+		readResp.Private = req.Private.Provider
+		resp.Private = req.Private
+	}
+
 	logging.FrameworkDebug(ctx, "Calling provider defined Resource Read")
 	resourceImpl.Read(ctx, readReq, &readResp)
 	logging.FrameworkDebug(ctx, "Called provider defined Resource Read")
 
 	resp.Diagnostics = readResp.Diagnostics
 	resp.NewState = &readResp.State
+
+	if readResp.Private != nil {
+		resp.Private.Provider = readResp.Private
```

Comment:
We may need to instantiate `resp.Private` when there was no `req.Private` -- it'd be great to have unit testing for that scenario.

Reply:
Good shout. Blows up with a nil pointer otherwise :grimacing: